dev/run-tests.py

```diff
@@ -0,0 +1,417 @@
+            " ".join(sbt_maven_profile_args + ["-Phive", "-Phive-thriftserver"])
+
+    # if we only have changes in SQL build a custom test string
+    if "SQL" in test_suite and "CORE" not in test_suite:
+        sbt_maven_test_args = ["catalyst/test",
+                               "sql/test",
+                               "hive/test", 
+                               "hive-thriftserver/test",
+                               "mllib/test"]
+    else:
+        sbt_maven_test_args = ["test"]
+
+    # get the latest sbt maven profile arguments
+    sbt_maven_profile_args = os.environ.get(SBT_MAVEN_PROFILE_ARGS_ENV).split()
+
+    print "[info] Running Spark tests with these arguments:",
+    print " ".join(sbt_maven_profile_args), 
+    print " ".join(sbt_maven_test_args)
+
+    if AMPLAB_JENKINS_BUILD_TOOL == "maven":
+        exec_maven(["test"] + sbt_maven_profile_args + ["--fail-at-end"])
+    else:
+        exec_sbt(sbt_maven_profile_args + sbt_maven_test_args)
+
+
+def run_python_tests(test_suite=[]):
+    set_title_and_block("Running PySpark tests", "BLOCK_PYSPARK_UNIT_TESTS")
+    
+    # Add path for Python3 in Jenkins if we're calling from a Jenkins machine
+    if AMPLAB_JENKINS:
+        os.environ["PATH"] = os.environ.get("PATH")+":/home/anaconda/envs/py3k/bin"
+
+    run_cmd(["./python/run-tests"])
+
+
+def run_sparkr_tests(test_suite=[]):
+    set_title_and_block("Running SparkR tests", "BLOCK_SPARKR_UNIT_TESTS")
+
+    if which("R"):
+        run_cmd(["./R/install-dev.sh"])
+        run_cmd(["./R/run-tests.sh"])
+    else:
+        print "Ignoring SparkR tests as R was not found in PATH"
+
+if __name__ == "__main__":
+    # Ensure the user home directory (HOME) is valid and is an absolute directory
+    if not USER_HOME_DIR or not os.path.isabs(USER_HOME_DIR):
+        print "[error] Cannot determine your home directory as an absolute path;",
+        print "ensure the $HOME environment variable is set properly."
+        sys.exit(1)
+
+    os.chdir(SPARK_PROJ_ROOT)
+
+    rm_r("./work")
+    rm_r(os.path.join(USER_HOME_DIR, ".ivy2/local/org.apache.spark"))
+    rm_r(os.path.join(USER_HOME_DIR, ".ivy2/cache/org.apache.spark"))
+
+    error_codes = get_error_codes("./dev/run-tests-codes.sh")
+
+    os.environ["CURRENT_BLOCK"] = error_codes["BLOCK_GENERAL"]
+
+    set_sbt_maven_profile_args()
+
+    java_exe = determine_java_executable()
+
+    if not java_exe:
+        print "[error] Cannot find a version of `java` on the system; please", 
+        print "install one and retry."
+        sys.exit(2)
+
+    java_version = determine_java_version(java_exe)
+
+    if java_version.minor < 8:
+        print "[warn] Java 8 tests will not run because JDK version is < 1.8."
+
+    test_suite = determine_test_suite()
+
+    run_apache_rat_checks()
+    
+    run_scala_style_checks()
+
+    run_python_style_checks()
+
+    build_apache_spark()
+
+    detect_binary_inop_with_mima()
+
+    run_scala_tests(test_suite)
```

Comment:
So there is a bunch of logic related to exactly which tests to build/run and it is scattered across a few different functions right now (`build_apache_spark`, `deteremine_test_suite`, `set_sbt_maven_profile_args`, and `run_scala_tests`). They use environment variables to communicate and it's pretty hard to reason about exactly what is run in which situations. I propose to refactor things slightly to make them easier to uderstand:

```
def get_hadoop_profiles():
  """ Return profiles indicating which Hadoop version to use."""
>> If AMPLAB_JENKINS_BUILD_PROFILE is set then add corresponding flags, otherwise add -Phadoop2.3 and -Pyarn

def get_build_profiles():
   """Returns a list of profiles to be used when building."""
>> Always return -Phive -Phive-thriftserver -Pkinesis-asl

def build_spark_maven():
def build_spark_sbt():
>> Runs sbt or maven package with the correct goals and build with get_hadoop_profiles() and get_build_profiles()

def determine_test_modules()
  """ Returns a string "sql-only" (only test Spark SQL), "all" (build everything)"""
   # This can be generalized later to return finer grained lists of modules

def run_scala_tests_maven()
>> This can support "all". I don't think it can support "sql-only" without more modification, so it's okay to fall back to "all" in that case.

def run_scala_tests_sbt()
>> This can support "all" and "sql-only" by modifying the test arguments.